### PR TITLE
Adding Validation Metrics for GPT/T5 prompt learning

### PR DESCRIPTION
**Pull request by Davood-M**
# What does this PR do ?

Adds the following metrics for GPT/T5 prompt learning:
- Accuracy -> exact string match
- BLEU score
- ROUGE-1, 2, 3, L

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [X] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [X] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py [hunk 1]
```diff
@@ -398,19 +408,19 @@ def validation_epoch_end(self, outputs):
...
+                    logging.info(f'Validation {metric}: {val}')
+                val_metric = list(val_metric_dict.items())[0][1]
+                metric_name = list(val_metric_dict.items())[0][0]
             else:
-                val_acc = torch.tensor(0.0).cuda()
+                val_metric = torch.tensor(0.0).cuda()
+                metric_name = 'Metric'
 
-            self.log('val_acc', val_acc, prog_bar=True, rank_zero_only=True)
+            self.log(f'Validation {metric_name}', val_metric, prog_bar=True, rank_zero_only=True)
```
Comment (arendu) [inline nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py:423]: minor, can we stick with val_{metric_name}?

File: nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py [hunk 2]
```diff
@@ -74,6 +75,15 @@ class MegatronT5PromptLearningModel(MegatronBasePromptLearningModel):
     def __init__(self, cfg: DictConfig, trainer: Trainer):
         super().__init__(cfg, trainer)
 
+        # define validation metric
```
Comment (arendu) [inline nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py:78]: Can this logic be moved to the base class?

File: nemo/collections/nlp/metrics/prompt_learning_metrics.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+                    scores['rouge1'].fmeasure,
+                    scores['rouge2'].fmeasure,
+                    scores['rouge3'].fmeasure,
+                    scores['rougeL'].fmeasure,
+                ]
+            )
+
+        all_rouges = np.mean(np.array(all_rouges), axis=0).tolist()
+
+        return {
```
Comment (arendu) [inline nemo/collections/nlp/metrics/prompt_learning_metrics.py:65]: Why does the result need to be copied to CUDA?

File: nemo/collections/nlp/metrics/prompt_learning_metrics.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+class AccuracyScore(object):
+    def get_score(self, ground_truth, predicted_text):
+        corrects = 0
+        for (pred, label) in zip(predicted_text, ground_truth):
+            if pred == label:
+                corrects += 1
+
+        val_acc = corrects / len(ground_truth)
+        val_acc = torch.tensor(val_acc).cuda()
+        return {'accuracy': torch.tensor(val_acc).cuda()}
```
Comment (arendu) [inline nemo/collections/nlp/metrics/prompt_learning_metrics.py:31]: Why does the result need to be copied to CUDA?

File: nemo/collections/nlp/metrics/prompt_learning_metrics.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+class BLEUScore(object):
+    def __init__(self):
+        self.scorer = BLEU()
+
+    def get_score(self, ground_truth, predicted_text):
+        return {
+            'bleu_score': torch.tensor(
+                self.scorer.corpus_score(predicted_text, [[i] for i in ground_truth],).score
+            ).cuda()
```
Comment (arendu) [inline nemo/collections/nlp/metrics/prompt_learning_metrics.py:42]: Why does the result need to be copied to CUDA?

File: nemo/collections/nlp/metrics/prompt_learning_metrics.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        }
+
+
+class ROUGEScores(object):
+    def __init__(self, use_stemmer=True):
```
Comment (Zhilin123) [inline nemo/collections/nlp/metrics/prompt_learning_metrics.py:47]: Is there any case in which we want use_stemmer=False? Otherwise, maybe we don't want to expose this.

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_prompt_learning_model.py [hunk 7]
```diff
@@ -344,7 +373,40 @@ def validation_step(self, batch, batch_idx):
         if loss_mean.item == 0.0:
             loss_mean = []
 
-        return loss_mean
+        if self.cfg.get('report_validation_metric', False):
+            preds_text, labels_text = [], []
+            input_ids, labels, loss_mask, position_ids, attention_mask, taskname_ids = batch
+            input_lenghts = torch.cuda.LongTensor(
```
Comment (Zhilin123) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_prompt_learning_model.py:379]: Not sure if this is correct since different items in a batch will have different input lengths but this code will have input length the same in the batch.

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_prompt_learning_model.py [hunk 8]
```diff
@@ -344,7 +373,40 @@ def validation_step(self, batch, batch_idx):
...
+            )
+
+            for pred, label in zip(res['token_ids'], labels):
+                additional_special_tokens_ids = []
+                if hasattr(self.tokenizer.tokenizer, "additional_special_tokens_ids"):
+                    additional_special_tokens_ids = self.tokenizer.tokenizer.additional_special_tokens_ids
+
+                pred = [id for id in pred if id not in additional_special_tokens_ids]
+                label = [id for id in label if id not in additional_special_tokens_ids]
+                pred = self.tokenizer.ids_to_text(pred)
```
Comment (Zhilin123) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_prompt_learning_model.py:399]: Maybe just `ids_to_text(pred, skip_special_tokens=True)`?

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_prompt_learning_model.py [hunk 9]
```diff
@@ -185,6 +188,32 @@ def init_model(self, cfg: DictConfig, trainer: Trainer):
         self.lowest_val_loss = None
         self.prompt_encoder = None
 
+        # default inference related params -> for evaluation metrics
+        self.length_params: LengthParam = {
```
Comment (Zhilin123) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_prompt_learning_model.py:192]: Do we always want to just generate 30 tokens? Feel weird if we are doing a long sequence task like email generation but this can increase speed quite a lot.

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_prompt_learning_model.py [hunk 10]
```diff
@@ -185,6 +188,32 @@ def init_model(self, cfg: DictConfig, trainer: Trainer):
...
+            "top_p": 0.9,
+            "repetition_penalty": 1.2,
+            "add_BOS": True,
+            "all_probs": False,
+            "compute_logprob": False,
+        }
+
+        # define validation metric
+        if self.cfg.get('report_validation_metric', False):
+            if self.cfg.get('validation_metric', 'accuracy') == 'accuracy':
```
Comment (Zhilin123) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_prompt_learning_model.py:210]: Might be a little more readable if 
```
validation_metric = self.cfg.get('validation_metric', 'accuracy')
if validation_metric = "accuracy":
  self.validation_metric = AccuracyScore()
elif validation_metric = "bleu":
  self.validation_metric = BLEUScore()
```

File: nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py (hunk 2 above)
Comment (Davood-M) [reply nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py]: moved to the upper class, but I get error for GPT model which I added it in the init_model function.

File: nemo/collections/nlp/metrics/prompt_learning_metrics.py [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+import torch
+from rouge_score import rouge_scorer
+from sacrebleu import BLEU
+
+
+class AccuracyScore(object):
```
[the remark this answers is not in the document]
Comment (Davood-M) [reply nemo/collections/nlp/metrics/prompt_learning_metrics.py]: Created an abstract class for the metrics and removed cuda calls which I'm not sure why they were there (they were part of T5's metrics and I carried them over)

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_prompt_learning_model.py (hunk 9 above)
Comment (Davood-M) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_prompt_learning_model.py]: The inference part in now configurable through the YAML file and accepts the same params as the inference config.

File: examples/nlp/language_modeling/conf/megatron_gpt_prompt_learning_config.yaml [hunk 12]
```diff
@@ -57,6 +57,8 @@ model:
   validation_global_batch_size: ${model.global_batch_size}
   validation_micro_batch_size: ${model.micro_batch_size}
   validation_drop_last: False
+  report_validation_metric: False
+  validation_metric: 'bleu'
```
Comment (Zhilin123) [inline examples/nlp/language_modeling/conf/megatron_gpt_prompt_learning_config.yaml:61]: Is there a reason why the default validation metric on gpt and t5 are not the same?

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_prompt_learning_model.py (hunk 7 above)
Comment (Zhilin123) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_prompt_learning_model.py:383]: Please check this again!

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_prompt_learning_model.py [hunk 13]
```diff
@@ -344,7 +377,34 @@ def validation_step(self, batch, batch_idx):
...
+                (input_ids, torch.cuda.LongTensor(input_lenghts)),
+                self.tokenizer,
+                self.length_params,
+                self.sampling_params,
+                task_ids=taskname_ids,
+            )
+
+            for pred, label in zip(res['token_ids'], labels):
+                pred = self.tokenizer.ids_to_text(pred)
+                label = self.tokenizer.ids_to_text(label)
```
Comment (Zhilin123) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_prompt_learning_model.py:398]: If you can, please leave a comment that mentions that ids_to_text strips special tokens by itself, in case, it's not obvious from the name itself.

**Review by Zhilin123 — COMMENTED** [review]
LGTM, some small questions, plus reiterating one point you might have missed

**Review by arendu — APPROVED** [review]
LGTM! would like to have small changes to the config vars (discussed offline).

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_prompt_learning_model.py [hunk 14]
```diff
@@ -353,7 +386,44 @@ def validation_step(self, batch, batch_idx):
...
 
-        return loss_mean
+        if self.cfg.get('report_validation_metric', False):
+            preds_text, labels_text = [], []
+            input_ids, labels, loss_mask, position_ids, attention_mask, taskname_ids = batch
+            input_lenghts = torch.argmax(loss_mask, 1, keepdim=True)
+
+            res = megatron_gpt_generate(
+                self.cuda(),
+                (
```
Comment (Zhilin123) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_prompt_learning_model.py:396]: Can we split this out into multiple lines for readability? Preferably also explain to people why we need to concatenate torch.zeros --> "it's because megatron_gpt_generate expects the input_ids to contain tokens for it to overwrite instead of appending to

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_prompt_learning_model.py [hunk 15]
```diff
@@ -353,7 +386,44 @@ def validation_step(self, batch, batch_idx):
         if loss_mean.item == 0.0:
             loss_mean = []
 
-        return loss_mean
+        if self.cfg.get('report_validation_metric', False):
+            preds_text, labels_text = [], []
+            input_ids, labels, loss_mask, position_ids, attention_mask, taskname_ids = batch
+            input_lenghts = torch.argmax(loss_mask, 1, keepdim=True)
```
Comment (Zhilin123) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_prompt_learning_model.py:392]: Explain in a comment what we are doing here since it's not too clear e.g. Finding the first non_zero element index for the answer to find out how many prompt tokens there are as megatron_gpt_generate expects this in input_ids.

Comment (Zhilin123) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_prompt_learning_model.py:392]: nit: can we change input_lenghts --> input_lengths. Technically both are ok but trying to be consistent with self.length_params

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_prompt_learning_model.py [hunk 16]
```diff
@@ -353,7 +386,44 @@ def validation_step(self, batch, batch_idx):
         if loss_mean.item == 0.0:
             loss_mean = []
 
-        return loss_mean
+        if self.cfg.get('report_validation_metric', False):
+            preds_text, labels_text = [], []
```
Comment (Zhilin123) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_prompt_learning_model.py:390]: Maybe declare these variables much lower (ie close to line 413) See https://softwareengineering.stackexchange.com/questions/113262/why-declare-variables-close-to-where-they-are-used

**Review by Zhilin123 — APPROVED** [review]
LGTM with minor stylistic comments. Please conducts the experiments before merging to see if we get reasonable metrics for tasks and understand the potential impact of additional time it will take.

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_prompt_learning_model.py [hunk 17]
```diff
@@ -353,7 +386,44 @@ def validation_step(self, batch, batch_idx):
...
+            input_ids, labels, loss_mask, position_ids, attention_mask, taskname_ids = batch
+            input_lenghts = torch.argmax(loss_mask, 1, keepdim=True)
+
+            res = megatron_gpt_generate(
+                self.cuda(),
+                (
+                    torch.cat(
+                        (
+                            input_ids,
+                            torch.zeros(
```
Comment (Zhilin123) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_prompt_learning_model.py:400]: One thing I realized is that this cannot be a torch.zeros. instead It has to a matrix filled with the pad_ids so something like torch.ones(...) * self.tokenizer.pad_id
